### PR TITLE
SCT-3141 Add support for config repo GitHub release workflow

### PR DESCRIPTION
**Pull request by dauglyon**
Adds support for downloading the validation config from a github release. Addresses step two of the [Implementation Plan for Sample Service Validator Workflow](https://docs.google.com/document/d/1DKN0rbQVaMYd88nmXt-TDIZF2CpkhFK-ewqz9n713TE/edit). This PR uses the github API to grab the assets. This is because the rate-limit (it seems only for some IPs) can be ~10 downloads per hour ([instructions to check your own rate limit](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#checking-your-rate-limit-status)). Unfortunately this seems undocumented, but I have hit it a few times. This goes up to the documented 1000 requets per hour per repo when using the API with a token.

### CODE REVIEW

File: lib/SampleService/core/config.py
Comment (slebras) [inline lib/SampleService/core/config.py:240]: should "metadata_validation.yml" perhaps be assignable?

**Review by slebras — COMMENTED** [review]
1 comment to clear up

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:240]: That is probably a good idea. I wasn't sure how many new config variables we wanted, but this makes sense for sure.

**Review by slebras — APPROVED** [review]
Looks good

File: deploy.cfg.tmpl
Comment (eapearson) [inline deploy.cfg.tmpl:74]: Env var typo? sb `metadata_validator_config_prerelease`

**Comment by eapearson** [comment]
The changes to, and the presence of, `deploy.cfg`, does seem superfluous. That file is generated from `deploy.cfg.tmpl` in the entrypoint script, so deploy.cfg will be overwritten.
It looks to be a remnant from when python was used to process the config file.

File: deploy.cfg.tmpl
Comment (eapearson) [inline deploy.cfg.tmpl:76]: do we want this to default to an empty string, or to fail? The token is required, with no sensible default.
same for the config url.
so, rather,
`{{ .Env.github_token }}`
etc.

Also, this change will require synchronized changes to wherever the deployment configs are kept - probably gitlab.

File: Pipfile
Comment (eapearson) [inline Pipfile:36]: this dependency should be pinned. 
it is typical to not pin typings (but that is perhaps debatable too), but regular libraries should be.

File: deploy.cfg.tmpl
Comment (dauglyon) [reply deploy.cfg.tmpl:74]:
```suggestion
metadata-validator-config-prerelease = {{ default .Env.metadata_validator_config_prerelease "0" }}
```

File: deploy.cfg.tmpl
Comment (dauglyon) [reply deploy.cfg.tmpl:76]: a good point

**Comment by MrCreosote** [comment]
The first thing I notice is that there's no coverage information...

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:225]: This line is too long...
```suggestion
def get_validators(
        repo_path: Optional[str] = None,
        repo_file: Optional[str] = None,
        prerelease_ok: bool = False,
        url: Optional[str] = None,
        token: Optional[str] = None
        ) -> MetadataValidatorSet:
```

Also there's no reason to make a bool optional unless you want trinary logic, which I don't think you want here

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:133]:
```suggestion
        prerelease_ok=(metaval_prelease_ok or '').lower() in ('true', 'yes', 'y', '1'),
```

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:134]: The `or None` is redundant, if there's no string `_check_string` returns `None`

```suggestion
        url=metaval_url,
```

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:135]: Same re `or None`

```suggestion
        token=github_token) if (
```

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:136]: IMO this would be a lot easier to read with a standard if block (this incorporates all the other suggestions above)

```suggestion
    if metaval_url or metaval_repo:
        metaval = get_validators(
            repo_path=metaval_repo,
            repo_file=metaval_filename,
            prerelease_ok=(metaval_prelease_ok or '').lower() in ('true', 'yes', 'y', '1'),
            url=metaval_url
            token=github_token
        )
    else:
        metaval = MetadataValidatorSet()
```

File: deploy.cfg.tmpl
Comment (MrCreosote) [inline deploy.cfg.tmpl:76]:
```suggestion
# Whether or not to consider prereleases when looking for the metadata validator config release asset.
```

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py:86]: That might not give the error message you want though, might be better to do a manual check

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py:136]: Also I'd handle the boolean checking in a specific method for booleans like `_check_bool` mentioned above

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:231]: Method documentation needs to be updated

File: Pipfile
Comment (MrCreosote) [inline Pipfile:36]: Needs release notes and possibly a version bump if it hasn't been bumped already

File: test/core/config_test.py
Comment (MrCreosote) [inline test/core/config_test.py:100]: There's no tests for the new validation download

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:246]:
```suggestion
            except Exception as e:
                raise RuntimeError(f'Fetching releases from repo {repo_path} failed.') from e
```

Otherwise debugging a fail is going to suck. I have seen this. I have done this. You do not want this

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:252]: Add the release tag / version to the error message

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:255]: Add the release tag / version and `repo_file` to the error message

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:267]: Add release & filename info if not present in the url

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:269]: If repo_path, include the other relevant info - release & filename info

File: lib/SampleService/core/config.py
Comment (MrCreosote) [inline lib/SampleService/core/config.py:272]: Same re relevant info for repo_path

**Review by MrCreosote — CHANGES_REQUESTED** [review]

File: Pipfile
Comment (dauglyon) [reply Pipfile:35]: I'm not sure what I do to change this? Does this repo have a coverage action?

Comment (MrCreosote) [reply Pipfile:35]: coveralls was running as of my last PR: https://github.com/kbase/sample_service/pull/380

What happened in the meantime isn't clear to me

Comment (dauglyon) [reply Pipfile:35]: This is just the a types lib needed for supporting pygithub. Used pygithub so I didn't have to handroll all the different github api calls / typing / error states. If we want to extend to supporting versions/tags that would be even more back&forth. Used the github api over raw links, as raw links are much more strictly rate-limited (I hit the rate limit just during dev).

Comment (MrCreosote) [reply Pipfile:35]: Sorry, I mean why is the PR writ large necessary, not this particular line. Ignore the line contents

> Used the github api over raw links, as raw links are much more strictly rate-limited (I hit the rate limit just during dev).

It seems this is the purpose of the PR though? What is the limit? I can't imagine starting the server enough (which is when the config gets pulled IIRC) could hit any reaonable rate limit

File: deploy.cfg.tmpl
Comment (MrCreosote) [reply deploy.cfg.tmpl:77]: I still feel a bit queasy about this... it seems like a really easy way to bork your environment by momentarily fat fingering a release somewhere else. I still think specifying the version would be the right way to go unless there's a good reason not to do that

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:252]: This error doesnt rely on a release version/tag as it is failing to fetch any releases at all.

Comment (dauglyon) [reply lib/SampleService/core/config.py:252]: Oh oops
my IDE had this comment pointing to the wrong line

File: test/core/config_test.py
Comment (dauglyon) [reply test/core/config_test.py:100]: This is going to be tricky as I'll need to mock the github api, but I can give it a go

Comment (MrCreosote) [reply test/core/config_test.py:100]: Usually I want self-contained tests, but for this case my initial thought would be to just set up a github repo to test against. Mocking all that stuff seems like too much work for too little gain over just relying on an external resource. Is there a reason I'm not seeing that that's not doable or a bad idea?

Comment (MrCreosote) [reply test/core/config_test.py:100]: ... AND ANOTHER THING... mocking external dependencies is an anti-pattern anyway since if the external APIs change your tests will continue to pass but your actual deployments will fail

File: deploy.cfg.tmpl
Comment (dauglyon) [reply deploy.cfg.tmpl]: added optional release tag config

Comment (MrCreosote) [reply deploy.cfg.tmpl]: :+1:

Comment (MrCreosote) [reply deploy.cfg.tmpl]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py:88]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py:261]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: Here just the repo_path was removed - I don't think that was the intent of the change?

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: Same as above. I don't think `config_asset == True` necessarily means that there's no repo path, right? Or maybe I don't quite grok the logic

File: test/core/config_test.py
Comment (MrCreosote) [reply test/core/config_test.py:100]: I don't see a test for a few cases:

* No url or repo supplied
* the no releases case
* the no assets case
* parser errors

File: Pipfile
Comment (MrCreosote) [reply Pipfile:36]: :+1:

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py]: It does mean that, yes. If a url is provided, it isnt loaded via a repo asset (config_asset) and doesn't use the repo_path logic (see lines 236-241)

File: Pipfile
Comment (dauglyon) [reply Pipfile:35]: More details on this in the new PR desc

Comment (MrCreosote) [reply Pipfile:35]: Wow that is a crazy low limit

Comment (MrCreosote) [reply Pipfile:35]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: never mind, see comment below

Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: test/core/config_test.py
Comment (dauglyon) [reply test/core/config_test.py:100]: For parser errors, they're just the same error one gets from the URL flow. I don't see good reason to go through the process of retesting that part. The others I can add cases for. Perhaps excluding the "no releases" case, thats something I'm less comfortable with testing without a mock because it's really easy to break that test if someone ends up creating a release in whatever repo is targeted.

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py]: it was yes, see my reply below

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: yep, you're right after reading from the top rather than skimming like a lazy chump

I kind of think the method is too long now (~60 lines) but I leave it up to you if you want to do anything about it

Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: test/core/config_test.py
Comment (MrCreosote) [reply test/core/config_test.py:100]: > For parser errors, they're just the same error one gets from the URL flow. 

maybe I'm misunderstanding again, but it looks like there's one branch for `config_asset==True` and one for `False`, and only the latter is tested?

Comment (MrCreosote) [reply test/core/config_test.py:100]: > Perhaps excluding the "no releases" case, thats something I'm less comfortable with testing without a mock because it's really easy to break that test if someone ends up creating a release in whatever repo is targeted.

make a new repo and archive it, and note in the readme it's for X test in Y repo, don't touch it?

Comment (MrCreosote) [reply test/core/config_test.py:100]: Oh, I missed a case for where the repo is provided but no asset filename

Comment (MrCreosote) [reply test/core/config_test.py:100]: coverage info would make this a lot simpler :P

Comment (MrCreosote) [reply test/core/config_test.py:100]: There's this org for test repos btw https://github.com/kbasetest

Comment (dauglyon) [reply test/core/config_test.py:100]: Actually the "no assets" case seems to never happen at the moment on GitHub as all releases include an asset file of the zipped source.

Comment (dauglyon) [reply test/core/config_test.py:100]: Unfortunately I have no privs for https://github.com/kbasetest

Comment (MrCreosote) [reply test/core/config_test.py:100]: > Actually the "no assets" case seems to never happen at the moment on GitHub as all releases include an asset file of the zipped source.

I'd ditch that check and add a comment mentioning that in that case

Comment (MrCreosote) [reply test/core/config_test.py:100]: > Unfortunately I have no privs

Invited

Comment (dauglyon) [reply test/core/config_test.py:100]: Added tests for check_bool, no releases, no URL or repo_path, and repo_path but no repo_asset. Unfortunately still have no idea why coverage isnt running on this PR but I dont see it on many other PRs in the repo...

Comment (MrCreosote) [reply test/core/config_test.py:100]: All I can say is that it ran on every PR I ever did from very early on, I think it was one of the first things I set up

File: Pipfile
Comment (MrCreosote) [reply Pipfile:35]: Per offline convos, we'll pass on fixing coverage in this PR but note it needs to happen soon, preferably before any more PRs get merged

Comment (MrCreosote) [reply Pipfile:35]: :+1:

File: test/core/arg_checkers_test.py
Comment (MrCreosote) [inline test/core/arg_checkers_test.py:5]:
```suggestion
from SampleService.core.arg_checkers import (
    check_string,
    check_bool,
    not_falsy,
    not_falsy_in_iterable
)
```

**Review by MrCreosote — CHANGES_REQUESTED** [review]

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:291]: test_config_get_validators_github_bad_tag covers a bad tag. Multiple tests don't provide an explicit tag

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:294]: what is "this"?

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:321]: Mentioned I think in a previous comment, I'm not of the opinion one is needed, as this is the same (already tested) bad-yaml case with different err text. Covered by test_config_get_validators_fail_bad_yaml

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:294]: If you mean the no-assets case. It's not currently replicable with GH, but the API spec doesnt specifically disallow an empty array here so, this is a "just in case" error that would require mocking GH reponses to test, which i think would be wild overkill

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:314]: Again, its just differentiated error text for an already tested error

File: test/core/arg_checkers_test.py
Comment (MrCreosote) [reply test/core/arg_checkers_test.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py]: :+1:

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py:291]: Hmm, I'm not seeing those. I only see one happy path test and it uses a tag

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py:294]: If it's practically impossible for this case to occur, I would just take code path out, add a comment that currently google will always have > 0 assets. If things change it'll throw a `TypeError` which I think is fine

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py:321]: I guess I don't understand why you wouldn't add a test? It's an uncovered code path and so should get a test for that code path IMO. The tests look pretty simple to write so naively it seems like a 5 minutes thing. I would certainly write one if it was my PR

File: lib/SampleService/core/config.py
Comment (MrCreosote) [reply lib/SampleService/core/config.py:314]: Same comment as below re test writing

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:321]: Mostly because it's a weird amount of overhead to use a third party service in tests to test a code path that is only differentiated by earlier (tested) behavior. It's also not a functionally different code path, it just changes what string is thrown...

File: lib/SampleService/core/config.py
Comment (dauglyon) [reply lib/SampleService/core/config.py:291]: bad tag: https://github.com/kbase/sample_service/pull/408/files#diff-daeaaa8b25848f3758165ca3a64ab5d0a9d57ace8942f4d66c6894665153da56R190
provides no tag and finds no tagged release:
https://github.com/kbase/sample_service/pull/408/files#diff-daeaaa8b25848f3758165ca3a64ab5d0a9d57ace8942f4d66c6894665153da56R218

**Pull request closed without merging**
